Mathlib/ModelTheory/Equivalence.lean

```diff
@@ -53,6 +56,52 @@ protected theorem trans {φ ψ θ : L.BoundedFormula α n} (h1 : φ ⟹[T] ψ) (
 
 instance : IsTrans (L.BoundedFormula α n) T.Imp := ⟨fun _ _ _ => Imp.trans⟩
 
+instance : IsPreorder (L.BoundedFormula α n) T.Imp where
+
+/-- Implication induces a preorder on formulas. -/
+def preorder (T : L.Theory) : Preorder (L.BoundedFormula α n) where
+  le := T.Imp
+  le_refl := Imp.refl
+  le_trans _ _ _ := Imp.trans
+
+theorem realize_bd_imp {φ ψ : L.BoundedFormula α n} (h : φ ⟹[T] ψ)
+    {v : α → M} {xs : Fin n → M} : φ.Realize v xs → ψ.Realize v xs :=
+  BoundedFormula.realize_imp.1 (h.realize_boundedFormula M)
+
+theorem realize_imp {φ ψ : L.Formula α} {M : Type*} [Nonempty M]
+    [L.Structure M] [M ⊨ T] (h : φ ⟹[T] ψ) {v : α → M} :
+    φ.Realize v → ψ.Realize v :=
+  h.realize_bd_imp
+
+theorem models_sentence_imp {φ ψ : L.Sentence} {M : Type*} [Nonempty M]
+    [L.Structure M] [M ⊨ T] (h : φ ⟹[T] ψ) :
+    M ⊨ φ → M ⊨ ψ :=
+  h.realize_imp
+
+protected theorem not {φ ψ : L.BoundedFormula α n} (h : φ ⟹[T] ψ) :
+    ψ.not ⟹[T] φ.not :=
+  fun _ _ _ h1 h2 => h1 (h.realize_bd_imp h2)
+
+protected theorem all {φ ψ : L.BoundedFormula α (n + 1)}
+    (h : φ ⟹[T] ψ) : φ.all ⟹[T] ψ.all :=
+  fun _ _ _ => forall_imp fun _ => h.realize_bd_imp
+
+protected theorem ex {φ ψ : L.BoundedFormula α (n + 1)} (h : φ ⟹[T] ψ) :
+    φ.ex ⟹[T] ψ.ex :=
+  h.not.all.not
+
+theorem imp_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : ψ ⟹[T] φ) (h' : φ' ⟹[T] ψ') :
+    (φ.imp φ') ⟹[T] (ψ.imp ψ') :=
+  fun _ _ _ h1 h2 => h'.realize_bd_imp (h1 (h.realize_bd_imp h2))
+
+theorem sup_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : φ ⟹[T] ψ) (h' : φ' ⟹[T] ψ') :
+    (φ ⊔ φ') ⟹[T] (ψ ⊔ ψ') :=
+  h.not.imp_mono h'
+
+theorem inf_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : φ ⟹[T] ψ) (h' : φ' ⟹[T] ψ') :
+    (φ ⊓ φ') ⟹[T] (ψ ⊓ ψ') :=
+  (h.imp_mono h'.not).not
```

Comment:
Does `gcongr` eat these kinds of lemmas?

Reply:
I'd not seen `gcongr` before - I've tried using it to no avail, but it might just be because I'm unfamiliar with it.

Reply:
What error do you get?

Reply:
Depending on which lemma, `gcongr failed, not a relation` or `gcongr did not make progress`

Reply:
Oh sorry I mean
```suggestion
@[gcongr]
theorem realize_imp {φ ψ : L.Formula α} {M : Type*} [Nonempty M]
    [L.Structure M] [M ⊨ T] (h : φ ⟹[T] ψ) {v : α → M} :
    φ.Realize v → ψ.Realize v :=
  h.realize_bd_imp

@[gcongr]
theorem models_sentence_imp {φ ψ : L.Sentence} {M : Type*} [Nonempty M]
    [L.Structure M] [M ⊨ T] (h : φ ⟹[T] ψ) :
    M ⊨ φ → M ⊨ ψ :=
  h.realize_imp

@[gcongr]
protected theorem not {φ ψ : L.BoundedFormula α n} (h : φ ⟹[T] ψ) :
    ψ.not ⟹[T] φ.not :=
  fun _ _ _ h1 h2 => h1 (h.realize_bd_imp h2)

@[gcongr]
protected theorem all {φ ψ : L.BoundedFormula α (n + 1)}
    (h : φ ⟹[T] ψ) : φ.all ⟹[T] ψ.all :=
  fun _ _ _ => forall_imp fun _ => h.realize_bd_imp

@[gcongr]
protected theorem ex {φ ψ : L.BoundedFormula α (n + 1)} (h : φ ⟹[T] ψ) :
    φ.ex ⟹[T] ψ.ex :=
  h.not.all.not

@[gcongr]
theorem imp_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : ψ ⟹[T] φ) (h' : φ' ⟹[T] ψ') :
    (φ.imp φ') ⟹[T] (ψ.imp ψ') :=
  fun _ _ _ h1 h2 => h'.realize_bd_imp (h1 (h.realize_bd_imp h2))

@[gcongr]
theorem sup_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : φ ⟹[T] ψ) (h' : φ' ⟹[T] ψ') :
    (φ ⊔ φ') ⟹[T] (ψ ⊔ ψ') :=
  h.not.imp_mono h'

@[gcongr]
theorem inf_mono {φ ψ φ' ψ' : L.BoundedFormula α n} (h : φ ⟹[T] ψ) (h' : φ' ⟹[T] ψ') :
    (φ ⊓ φ') ⟹[T] (ψ ⊓ ψ') :=
  (h.imp_mono h'.not).not
```

Reply:
Ah - ok. I've added the `gcongr` tag to all the lemmas that will take it.